src/Sarif/Core/ToolComponent.cs

```diff
@@ -1,13 +1,60 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
 namespace Microsoft.CodeAnalysis.Sarif
 {
     /// <summary>
     /// A component of the analysis tool that was run, such as the primary driver or a plug-in.
     /// </summary>
     public partial class ToolComponent
     {
+        private Dictionary<string, ReportingDescriptor> _cachedRulesById;
+        private Dictionary<string, ReportingDescriptor> _cachedRulesByGuid;
+
+        private void BuildRuleCaches()
+        {
+            _cachedRulesById = new Dictionary<string, ReportingDescriptor>();
+            _cachedRulesByGuid = new Dictionary<string, ReportingDescriptor>();
+
+            foreach (ReportingDescriptor r in this.Rules ?? Enumerable.Empty<ReportingDescriptor>())
+            {
+                if (r.Id != null) { _cachedRulesById[r.Id] = r; }
+                if (r.Guid != null) { _cachedRulesByGuid[r.Guid] = r; }
+            }
+        }
+
+        public ReportingDescriptor GetRuleById(string ruleId)
+        {
+            ReportingDescriptor rule = null;
+
+            // Build lookup if not built or possibly out-of-date
+            if (_cachedRulesById == null || !_cachedRulesById.TryGetValue(ruleId, out rule))
```

Comment:
>) [](start = 91, length = 1)

Nits:

1. you don't have to predeclare `rule`: `TryGetValue(ruleId, out ReportingDescriptor rule)`
2. you could write `if (_cachedRulesById?.TryGetValue(ruleId, out ReportingDescriptorRule) != true)`

Reply:
I agree with #2 but I don't like inline declaration when the variable will be used in a broader context than the statement. Since this one isn't just for the if true case, I think it's weird to declare it inline, technically in an expression that won't execute (if _cachedRulesById was null) and then use it outside the whole block. I know the compiler handles it fine, but it looks weird to me.

Reply:
Though actually I don't love #2 either. You have to think carefully about how the null boolean evaluates to the condition, whereas a X != null && .. condition is very trivial to understand.